### PR TITLE
fix lock table map to 1.1

### PR DESCRIPTION
**Pull request by zhangxu19830126**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #14146 

## What this PR does / why we need it:
Fix same lock table create muti-times in handle bind changed.

### CODE REVIEW

**Comment by matrix-meow** [comment]
@zhangxu19830126 Thanks for your contributions!

**Here are review comments for file pkg/lockservice/lock_table_allocator_test.go:**
Title: fix lock table map to 1.1

Body: This pull request fixes the issue #14146. The problem was that the lock table was being created multiple times in the handle bind changed. This PR addresses this issue by making changes to the file pkg/lockservice/lock_table_allocator_test.go.

Changes Made in pkg/lockservice/lock_table_allocator_test.go:
- Removed the import of the "sync" package.
- Replaced the creation of a sync.Map with the creation of a lockTableHolder struct.
- Replaced the use of the sync.Map.Store() method with the use of the lockTableHolder.set() method.

Review:

1. In the pull request title, it would be helpful to provide more context about the specific fix being made. For example, instead of "fix lock table map to 1.1", a more descriptive title like "Fix issue with multiple lock table creation in handle bind" would be more informative.

2. In the pull request body, it would be beneficial to provide a brief explanation of the issue being fixed and why it is important to address it. This would help reviewers understand the significance of the changes being made.

3. In the changes made to pkg/lockservice/lock_table_allocator_test.go, there are a few areas that could be improved:

- It is unclear why the import of the "sync" package was removed. If it is no longer needed, it would be helpful to provide a comment explaining why.

- The replacement of the sync.Map with the lockTableHolder struct is a good improvement. However, it would be helpful to provide more information about the lockTableHolder struct and its purpose. Additionally, it would be beneficial to include comments explaining the purpose of the fields and methods in the struct.

- The replacement of the sync.Map.Store() method with the lockTableHolder.set() method is a good improvement. However, it would be helpful to include comments explaining the purpose of the set() method and how it differs from the Store() method.

4. It is unclear if there are any security issues addressed in this pull request. If there are any security concerns related to the fix being made, it would be important to highlight them and provide suggestions for addressing them.

Suggestions for Optimization:

- Consider providing more descriptive and informative commit messages for each individual change made in the pull request. This would make it easier for reviewers to understand the purpose and impact of each change.

- Consider adding comments throughout the code to explain the purpose and functionality of different sections. This would make the code more readable and maintainable.

- Consider running tests to ensure that the changes made in the pull request do not introduce any new issues or regressions.

Overall, the changes made in this pull request address the issue of multiple lock table creation in the handle bind. However, there are areas that could be improved in terms of clarity and documentation. Additionally, it would be helpful to provide more context about the issue being fixed and any potential security concerns.

**Here are review comments for file pkg/lockservice/lock_table_keeper.go:**
Review:

Title: fix lock table map to 1.1

Body: This pull request fixes the issue #14146, which is related to creating the same lock table multiple times in handle bind changed.

Changes made in `pkg/lockservice/lock_table_keeper.go`:
- Removed the import of the `sync` package.
- Replaced the `tables` variable type from `sync.Map` to `lockTableHolder`.
- Updated the `NewLockTableKeeper` function signature to accept `tables` of type `*lockTableHolder`.
- Updated the `doKeepRemoteLock` function to use the `iter` method of `lockTableHolder` instead of `Range` method of `sync.Map`.
- Updated the `doKeepLockTableBind` function to use the `iter` and `removeWithFilter` methods of `lockTableHolder` instead of `Range` and `Delete` methods of `sync.Map`.

Overall, the changes made in this pull request seem to address the issue mentioned in the body. However, there are a few areas that could be improved:

1. Security: It's important to ensure that the changes made in this pull request do not introduce any security vulnerabilities. It would be beneficial to conduct a security review of the code changes to identify any potential security issues.

2. Code Optimization: The `doKeepRemoteLock` and `doKeepLockTableBind` functions can be further optimized. Instead of iterating over all the elements in the `tables` map, it would be more efficient to only iterate over the necessary elements. This can be achieved by maintaining a separate list of keys that need to be processed and iterating over that list instead.

3. Error Handling: The code changes do not include any error handling. It's important to handle any potential errors that may occur during the execution of the code, such as network errors or failures in accessing the `tables` map.

4. Documentation: It would be helpful to provide more detailed comments or documentation for the functions and variables that have been modified. This will make it easier for other developers to understand the purpose and functionality of the code.

To address these suggestions, the following actions can be taken:

1. Conduct a security review of the code changes to ensure there are no security vulnerabilities introduced.

2. Optimize the `doKeepRemoteLock` and `doKeepLockTableBind` functions by maintaining a separate list of keys that need to be processed and iterating over that list instead of iterating over all elements in the `tables` map.

3. Implement error handling for potential errors that may occur during the execution of the code, such as network errors or failures in accessing the `tables` map.

4. Add detailed comments or documentation for the modified functions and variables to improve code readability and understanding.

By addressing these suggestions, the code changes in this pull request can be further improved in terms of security, efficiency, and maintainability.

**Here are review comments for file pkg/lockservice/lock_table_keeper_test.go:**
Review:

Title: fix lock table map to 1.1

Body: This pull request fixes the issue #14146. The problem was that the lock table was being created multiple times in the handle bind changed. This PR addresses this issue by making changes to the file `pkg/lockservice/lock_table_keeper_test.go`.

Changes made in `pkg/lockservice/lock_table_keeper_test.go`:
- Line 16: Removed the import statement for the `sync` package.
- Line 62: Replaced the creation of a `sync.Map` with the creation of a `lockTableHolder` struct with an `id` field set to "s1" and a `tables` field set to an empty map of `uint64` to `lockTable`.
- Line 63: Replaced the `m.Store(0, ...)` with `m.set(0, ...)`.
- Line 69: Replaced the `m.Store(1, ...)` with `m.set(1, ...)`.
- Line 122: Replaced the `m.Store(1, ...)` with `m.set(1, ...)`.
- Line 128: Replaced the `m.Store(2, ...)` with `m.set(2, ...)`.
- Line 134: Replaced the `m.Store(3, ...)` with `m.set(3, ...)`.
- Line 144: Replaced the `m.Range(...)` with `m.iter(...)` and made corresponding changes to the function signature.
- Line 147: Replaced the `m.Load(3)` with `m.get(3)` and added an assertion to check if the value is not nil.

Problems:
1. Security Issue: The code does not appear to have any security issues.

Suggestions:
1. Optimization: The changes made in the pull request seem to be focused on fixing the issue. However, there is an opportunity to optimize the code further. Instead of using a `lockTableHolder` struct with a map, it might be more efficient to use a simple map directly. This would eliminate the need for the `lockTableHolder` struct and simplify the code.

Overall, the changes made in the pull request address the issue reported and do not introduce any security issues. However, there is an opportunity to optimize the code further by simplifying the data structure used.

**Here are review comments for file pkg/lockservice/service.go:**
Review:

Title: fix lock table map to 1.1

Body: The pull request fixes an issue where the same lock table is created multiple times in the handle bind changed function. It references issue #14146.

Changes made in the pull request:
- In the `service` struct, the `tables` field is changed from a `sync.Map` to a `lockTableHolder` pointer.
- In the `NewLockService` function, the `tables` field is initialized with a new `lockTableHolder` instance.
- In the `Close` method, the `tables` map is cleared using the `removeWithFilter` method of the `lockTableHolder` struct.
- In the `waitLockTableBind` method, the `loadLockTable` method is replaced with the `get` method of the `lockTableHolder` struct.
- In the `getLockTableWithCreate` method, the `loadLockTable` method is replaced with the `get` method of the `lockTableHolder` struct.
- In the `getLockTableWithCreate` method, the `createLockTableByBind` method is called and the result is stored in the `tables` map using the `set` method of the `lockTableHolder` struct.
- In the `handleBindChanged` method, the `tables` map is updated with the new lock table using the `set` method of the `lockTableHolder` struct.
- A new `lockTableHolder` struct is added, which replaces the functionality of the `sync.Map` for managing the lock tables.
- The `loadLockTable` method is removed from the `service` struct.
- The `lockTableHolder` struct is added with methods for getting, setting, iterating, and removing lock tables.

Potential problems and suggestions:
1. In the `service` struct, the `tables` field is changed from a `sync.Map` to a `lockTableHolder` pointer. This change may introduce potential concurrency issues if not handled correctly. It is important to ensure that proper locking mechanisms are used when accessing and modifying the `tables` field to avoid race conditions. Consider using `sync.RWMutex` to protect concurrent access to the `tables` field.

2. In the `Close` method, the `tables` map is cleared using the `removeWithFilter` method of the `lockTableHolder` struct. However, the filter function passed to `removeWithFilter` always returns `true`, which means that all lock tables will be removed. This may not be the intended behavior. Consider modifying the filter function to only remove specific lock tables based on certain conditions.

3. In the `getLockTableWithCreate` method, the `createLockTableByBind` method is called and the result is stored in the `tables` map using the `set` method of the `lockTableHolder` struct. However, there is no check to ensure that the lock table is successfully created before storing it in the map. Consider adding error handling to handle cases where the lock table creation fails.

4. In the `handleBindChanged` method, the `tables` map is updated with the new lock table using the `set` method of the `lockTableHolder` struct. However, there is no check to ensure that the new lock table is different from the existing one before updating the map. This may result in unnecessary updates to the map. Consider adding a check to compare the new and existing lock tables before updating the map.

5. In the `lockTableHolder` struct, the `iter` method is defined but not used in the pull request. Consider removing the unused method to improve code readability and maintainability.

Optimization suggestions:
1. Consider using a more descriptive variable name instead of `m` in the `lockTableHolder` struct methods to improve code readability.

2. In the `set` method of the `lockTableHolder` struct, the `old` variable is declared but not used. Consider removing the unused variable to improve code clarity.

3. In the `set` method of the `lockTableHolder` struct, the `ok` variable is used to check if the lock table already exists in the map. However, the `ok` variable is not used anywhere else in the method. Consider removing the unused variable to improve code readability.

4. In the `set` method of the `lockTableHolder` struct, the `newBind` variable is used to get the bind information from the new lock table. However, the `newBind` variable is not used anywhere else in the method. Consider removing the unused variable to improve code clarity.

Overall, the changes made in the pull request address the issue of creating the same lock table multiple times. However, there are potential concurrency issues and some areas that could be optimized. It is important to carefully review and test the changes to ensure they function as intended and do not introduce any new issues.

**Here are review comments for file pkg/lockservice/service_observability.go:**
Review:

Title: fix lock table map to 1.1

Body: The pull request fixes the issue #14146, which is related to creating the same lock table multiple times in handle bind changed.

Changes in pkg/lockservice/service_observability.go:
- In the `ForceRefreshLockTableBinds` function, the `contains` function now takes an additional parameter `_ lockTable`. However, this parameter is not used in the function body. It seems like a mistake and should be removed.
- The `ForceRefreshLockTableBinds` function has been modified to use the `removeWithFilter` method on the `s.tables` map instead of manually iterating over the map and deleting entries. This change improves the code readability and reduces the chances of introducing bugs.
- In the `IterLocks` function, the `s.tables.Range` method has been replaced with the `s.tables.iter` method. This change improves the code readability and reduces the chances of introducing bugs.

Overall, the changes in this pull request seem to address the issue and improve the code. However, there is a small mistake in the `contains` function parameter that should be fixed. Additionally, it would be helpful to provide more context or explanation in the pull request body to help reviewers understand the issue and the changes made.

**Here are review comments for file pkg/lockservice/service_remote.go:**
Review:

Title: fix lock table map to 1.1

Body: 
- The PR is a bug fix.
- It fixes issue #14146.
- The PR fixes the problem of creating the same lock table multiple times when the handle bind is changed.

Changes in pkg/lockservice/service_remote.go:
- Line 45: The change removes the address-of operator (&) before s.tables in the function call.
- The change fixes the issue of passing a pointer to s.tables instead of passing the value of s.tables.

Suggestions:
- The title and body of the pull request provide a clear description of the problem being fixed and the purpose of the PR.
- The changes made in the code appear to address the issue correctly.
- There are no security issues identified in the pull request.
- The changes made in the pull request are minimal and seem to be optimized for fixing the specific issue.

Overall, the pull request seems to be in good shape and can be merged after a final review.

**Here are review comments for file pkg/lockservice/service_remote_test.go:**
Title: fix lock table map to 1.1

Body: This pull request fixes the issue #14146. The issue was that the lock table was being created multiple times in the handle bind changed.

Changes in pkg/lockservice/service_remote_test.go:
- In the `TestLockWithBindTimeout` function, the code has been modified to use the `get` method of the `l2.tables` map instead of using `Load` and type assertion.
- In the `TestUnlockWithBindTimeout` function, the code has been modified to use the `get` method of the `l2.tables` map instead of using `Load` and type assertion.
- In the `TestGetLockWithBindTimeout` function, the code has been modified to use the `get` method of the `l2.tables` map instead of using `Load` and type assertion.
- In the `TestIssue12554` function, the code has been modified to use the `removeWithFilter` and `set` methods of the `l1.tables` map instead of using `Delete` and `Store`.
- The `checkBind` function has been modified to use the `get` method of the `s.tables` map instead of using `Load` and type assertion.

Overall, the changes made in this pull request seem to address the issue mentioned in the body. However, there are a few areas that could be improved and potential security issues to consider.

1. In the `TestLockWithBindTimeout`, `TestUnlockWithBindTimeout`, and `TestGetLockWithBindTimeout` functions, the code is using the `get` method of the `l2.tables` map without checking if the returned value is nil. This could potentially lead to a nil pointer dereference error if the value is not found in the map. It would be safer to check if the value is nil before accessing its properties.

Suggestion: Add a nil check before accessing the properties of the returned value in the `get` method.

2. In the `TestIssue12554` function, the code is using the `removeWithFilter` method of the `l1.tables` map without providing a filter function. This could potentially remove all entries from the map unintentionally. It would be safer to provide a filter function that checks if the entry matches the condition before removing it.

Suggestion: Provide a filter function in the `removeWithFilter` method to ensure only the desired entries are removed.

3. In the `checkBind` function, the code is using the `get` method of the `s.tables` map without checking if the returned value is nil. This could potentially lead to a nil pointer dereference error if the value is not found in the map. It would be safer to check if the value is nil before accessing its properties.

Suggestion: Add a nil check before accessing the properties of the returned value in the `get` method.

4. It would be helpful to add comments explaining the purpose and functionality of the modified code sections, especially in the `TestIssue12554` function.

Optimization suggestions:
- Consider using more descriptive variable names to improve code readability.
- Remove unused variables or code blocks to reduce clutter and improve maintainability.
- Consider refactoring repetitive code sections to reduce duplication and improve code maintainability.

Security considerations:
- Ensure that the changes made in this pull request do not introduce any security vulnerabilities or weaken the existing security measures. Review the code changes carefully to identify any potential security issues, such as improper handling of sensitive data or lack of input validation.

**Here are review comments for file pkg/lockservice/service_test.go:**
Review:

Title: fix lock table map to 1.1

Body: 
- The PR is marked as a bug fix.
- It fixes issue #14146.
- The PR fixes the problem of creating the same lock table multiple times when the handle bind is changed.

Changes in pkg/lockservice/service_test.go:
- Line 1350: The code previously used `s.tables.Store(table, new)` to store a new local lock table. It has been changed to `s.tables.set(table, new)`.

Feedback:
1. The title of the pull request is not very descriptive. It would be helpful to have a more specific title that clearly indicates what the fix is addressing.

2. The body of the pull request is concise and provides the necessary information about the problem being fixed and the issue it addresses.

3. The changes made in the code seem appropriate for fixing the issue. However, it would be helpful to have more context about the `s.tables` variable and the `set` method being used. Without this context, it is difficult to fully understand the impact of the change.

4. It is important to ensure that the fix does not introduce any security vulnerabilities. However, based on the provided information, it is not possible to determine if there are any security issues in the code.

Suggestions:
1. Provide a more descriptive title for the pull request that clearly indicates the problem being fixed.

2. Add more context about the `s.tables` variable and the `set` method being used in the code changes. This will help reviewers understand the impact of the change and ensure that it is correct.

3. Consider adding test cases to verify the fix and prevent regressions in the future.

4. If there are any security concerns related to the code changes, they should be addressed and documented in the pull request.

**Review by sukki37 — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Pull request merged**
